libraries/persistent_store/src/store.rs

```diff
@@ -128,3 +168,1089 @@ impl StoreUpdate {
+        self.write_slice(tail, &entry[..(footer * word_size) as usize])?;
+        self.write_slice(tail + footer, &entry[(footer * word_size) as usize..])?;
+        self.insert_init(tail, footer, key)
+    }
+
+    /// Removes an entry given its key.
+    ///
+    /// This is not an error if there is no entry for this key.
+    pub fn remove(&mut self, key: usize) -> StoreResult<()> {
+        let key = usize_to_nat(key);
+        if key > self.format.max_key() {
+            return Err(StoreError::InvalidArgument);
+        }
+        self.delete_keys(&[key], self.tail()?)
+    }
+
+    /// Removes an entry given a handle.
+    pub fn remove_handle(&mut self, handle: &StoreHandle) -> StoreResult<()> {
+        self.check_handle(handle)?;
+        self.delete_pos(handle.pos, self.format.bytes_to_words(handle.len))
+    }
+
+    /// Returns the maximum length in bytes of a value.
+    pub fn max_value_length(&self) -> usize {
+        self.format.max_value_len() as usize
+    }
+
+    /// Returns the value of an entry given its handle.
+    fn get_value(&self, handle: &StoreHandle) -> StoreResult<Vec<u8>> {
+        self.check_handle(handle)?;
+        let mut pos = handle.pos;
+        match self.parse_entry(&mut pos)? {
+            ParsedEntry::User(header) => {
+                let mut result = self.read_slice(handle.pos + 1, header.length);
+                if header.flipped {
+                    let last_byte = result.len() - 1;
+                    result[last_byte] = 0xff;
+                }
+                Ok(result)
+            }
+            ParsedEntry::Padding => Err(StoreError::InvalidArgument),
+            _ => Err(StoreError::InvalidStorage),
+        }
+    }
+
+    /// Initializes the storage if completely erased or partially initialized.
+    fn recover_initialize(&mut self) -> StoreResult<()> {
+        let word_size = self.format.word_size();
+        for page in 0..self.format.num_pages() {
+            let (init, rest) = self.read_page(page).split_at(word_size as usize);
+            if (page > 0 && !is_erased(init)) || !is_erased(rest) {
+                return Ok(());
+            }
+        }
+        let index = self.format.index_init(0);
+        let init_info = self.format.build_init(InitInfo {
+            cycle: 0,
+            prefix: 0,
+        });
+        self.storage_write_slice(index, &init_info)
+    }
+
+    /// Recovers a possible compaction interrupted while erasing the page.
+    fn recover_erase(&mut self) -> StoreResult<()> {
+        let mut pos = self.get_extremum_page_head(Ordering::Greater)?;
+        let end = pos.next_page(&self.format);
+        while pos < end {
+            let entry_pos = pos;
+            match self.parse_entry(&mut pos)? {
+                ParsedEntry::Internal(InternalEntry::Erase { .. }) => {
+                    return self.compact_erase(entry_pos)
+                }
+                ParsedEntry::Padding | ParsedEntry::User(_) => (),
+                _ => break,
```

Comment:
Yes this is intended, the store format invariant implies that there is at most one internal entry (marker, erase, clear) in the whole storage (the remove internal entries only appear after a marker).

The order in which we recover stuff is important. We first recover compaction (because otherwise the virtual storage is not well defined), then entry-level operations (now that the virtual storage is defined). And for compaction recovery we do it in opposite order as compaction would occur (to detect and continue from the latest change).